resteasy-jaxrs/src/main/java/org/jboss/resteasy/core/StringParameterInjector.java

```diff
@@ -80,40 +80,39 @@ protected void initialize(Class type, Type genericType, String paramName, Class
       baseType = type;
       baseGenericType = genericType;
 
-      if (type.isArray()) baseType = type.getComponentType();
-      if (List.class.isAssignableFrom(type))
-      {
-         isCollection = true;
-         collectionType = ArrayList.class;
-      }
-      else if (SortedSet.class.isAssignableFrom(type))
-      {
-         isCollection = true;
-         collectionType = TreeSet.class;
-      }
-      else if (Set.class.isAssignableFrom(type))
+      if (type.isArray())
       {
-         isCollection = true;
-         collectionType = HashSet.class;
+    	  baseType = type.getComponentType();
       }
-      if (isCollection)
+     
+      if (!baseType.isPrimitive())
       {
-         if (genericType != null && genericType instanceof ParameterizedType)
+         paramConverter = factory.getParamConverter(baseType, baseGenericType, annotations);
+         if (paramConverter != null) 
          {
-            ParameterizedType zType = (ParameterizedType) genericType;
-            baseType = Types.getRawType(zType.getActualTypeArguments()[0]);
-            baseGenericType = zType.getActualTypeArguments()[0];
+        	 return;
          }
-         else
-         {
-            baseType = String.class;
-            baseGenericType = null;
+         
+         collectionType = convertParameterTypeToCollectionType();
+         if(collectionType != null){
+        	 isCollection = true;
+	         if (genericType != null && genericType instanceof ParameterizedType)
+	         {
+	            ParameterizedType zType = (ParameterizedType) baseGenericType;
+	            baseType = Types.getRawType(zType.getActualTypeArguments()[0]);
+	            baseGenericType = zType.getActualTypeArguments()[0];
+	         }
+	         else
+	         {
+	            baseType = String.class;
+	            baseGenericType = null;
+	         }
+	         paramConverter = factory.getParamConverter(baseType, baseGenericType, annotations);
+	         if (paramConverter != null) 
+	         {
+	        	 return;
+	         }
          }
-      }
-      if (!baseType.isPrimitive())
-      {
-         paramConverter = factory.getParamConverter(baseType, baseGenericType, annotations);
-         if (paramConverter != null) return;
 
          unmarshaller = factory.createStringParameterUnmarshaller(baseType);

```

Comment:
Yes actually the same "double" treatment could be done for all calls from line 17 to the end. It will allow users to provide a way to convert `String` type to any type subclassing `List`, `Set ` or `SortedSet` before fallback on default implementation  `ArrayList`, `TreeSet ` and `HashSet ` and focusing on data held by the collection. 
The  "first" treatement will  end if one of the following is true :
> `paramConverter != null`
> `unmarshaller != null`
> `converter != null`
> `delegate != null` in case of `paramType.equals(HeaderParam.class)`
> `valueOf != null`